### PR TITLE
Fixed bug of elementwise ops

### PR DESCRIPTION
**Pull request by chenfeiyue-cfy**
Fixed bug that elementwise ops not keep the quantize params of the second input

Refine broadcast shape to avoid unnecessary permute around element wise ops

Type: Bug fix
Signed-off-by: Feiyue Chen [Feiyue.Chen@verisilicon.com](mailto:Feiyue.Chen@verisilicon.com)

### CODE REVIEW

File: op_map.cc [hunk 1]
```diff
@@ -657,30 +563,45 @@ struct SimpleOpWithFusedActiovationMapper
...
     TFLITE_LOG(TFLITE_LOG_INFO, "Creating %s op", name_.c_str());
 
-    bool boradcast_required = (inputs[0]->GetShape() != inputs[1]->GetShape());
+    bool boradcast_required = (inputs[0]->GetShape().size() != inputs[1]->GetShape().size());
     std::vector<std::shared_ptr<tim::vx::Tensor>> elementwise_inputs;
     if (boradcast_required) {
-      tim::vx::TensorSpec spec = inputs[0]->GetSpec();
-      spec = spec.AsTransientSpec();
-      auto broadcast_out = delegate->GetGraph()->CreateTensor(spec);
       if (inputs[0]->GetShape().size() > inputs[1]->GetShape().size()) {
```
Comment (sunshinemyson) [inline op_map.cc:569]: can we make this code more simple without the if-else ?

File: op_map.cc [hunk 2]
```diff
@@ -700,6 +621,82 @@ struct SimpleOpWithFusedActiovationMapper
...
+  bool HandleMapOp(vx::delegate::Delegate* delegate,
+                   std::vector<std::shared_ptr<tim::vx::Tensor>>& inputs,
+                   std::vector<std::shared_ptr<tim::vx::Tensor>>& outputs,
+                   const void* params) override {
+    TFLITE_LOG(TFLITE_LOG_INFO, "Creating %s op", name_.c_str());
+
+    bool boradcast_required = (inputs[0]->GetShape().size() != inputs[1]->GetShape().size());
+    std::vector<std::shared_ptr<tim::vx::Tensor>> elementwise_inputs;
+    if (boradcast_required) {
+      if (inputs[0]->GetShape().size() > inputs[1]->GetShape().size()) {
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply op_map.cc]: Done, I added it as a helper function in OpMapperBase

File: op_map.cc [hunk 3]
```diff
@@ -391,6 +391,17 @@ struct OpMapperBase : public vx::op_map::IOpMapper {
       const void* params) {
     return false;
   }
+
+  std::vector<int32_t> Extend_Broadcast(std::shared_ptr<tim::vx::Tensor>& longer_tensor, std::shared_ptr<tim::vx::Tensor>& shorter_tensor){
```
Comment (sunshinemyson) [inline op_map.cc:395]: const reference for parameter

File: op_map.cc [hunk 4]
```diff
@@ -657,41 +574,69 @@ struct SimpleOpWithFusedActiovationMapper
...
-      if (inputs[0]->GetShape().size() > inputs[1]->GetShape().size()) {
-        std::vector<uint32_t> shape = inputs[0]->GetShape();
-        std::vector<int32> broadcast_param;
-        for (auto iter = shape.begin(); iter != shape.end(); iter++) {
-          broadcast_param.push_back(*iter);
-        }
-        auto op_broadcast =
+
+    if (broadcast_required) {
+      int long_pointer = inputs[0]->GetShape().size() >
```
Comment (sunshinemyson) [inline op_map.cc:581]: naming suggestion: long_pointer to base_shape_idx

File: op_map.cc [hunk 5]
```diff
@@ -657,41 +574,69 @@ struct SimpleOpWithFusedActiovationMapper
...
+                   std::vector<std::shared_ptr<tim::vx::Tensor>>& inputs,
+                   std::vector<std::shared_ptr<tim::vx::Tensor>>& outputs,
+                   const void* params) override {
+    TFLITE_LOG(TFLITE_LOG_INFO, "Creating %s op", name_.c_str());
+
+    bool broadcast_required = (inputs[0]->GetShape().size() != inputs[1]->GetShape().size());
+    std::vector<std::shared_ptr<tim::vx::Tensor>> elementwise_inputs;
+
+    if (broadcast_required) {
+      int long_pointer = inputs[0]->GetShape().size() >
```
Comment (sunshinemyson) [inline op_map.cc:623]: can we reuse code in line 581?

File: op_map.cc [hunk 6]
```diff
@@ -391,6 +391,46 @@ struct OpMapperBase : public vx::op_map::IOpMapper {
       const void* params) {
     return false;
   }
+
+  std::vector<int32_t> Extend_Broadcast(const std::shared_ptr<tim::vx::Tensor>& base_shape_tensor,
```
Comment (sunshinemyson) [inline op_map.cc:395]: name conversion defeat:: function name should not with "_"

Comment (chenfeiyue-cfy) [reply op_map.cc]: Renamed

**Review by sunshinemyson — APPROVED** [review]

**Pull request merged**